news-bundle/src/EventListener/NewsFeedListener.php

```diff
@@ -0,0 +1,213 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\NewsBundle\EventListener;
+
+use Contao\ContentModel;
+use Contao\Controller;
+use Contao\CoreBundle\Cache\EntityCacheTags;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Image\ImageFactoryInterface;
+use Contao\CoreBundle\InsertTag\InsertTagParser;
+use Contao\Environment;
+use Contao\File;
+use Contao\FilesModel;
+use Contao\News;
+use Contao\NewsBundle\Event\FetchArticlesForFeedEvent;
+use Contao\NewsBundle\Event\TransformArticleForFeedEvent;
+use Contao\NewsModel;
+use Contao\StringUtil;
+use Contao\UserModel;
+use FeedIo\Feed\Item;
+use FeedIo\Feed\Item\Author;
+use FeedIo\Feed\Item\AuthorInterface;
+use FeedIo\Feed\Item\Media;
+use FeedIo\Feed\ItemInterface;
+use Symfony\Component\Filesystem\Path;
+
+class NewsFeedListener
+{
+    public function __construct(private readonly ContaoFramework $framework, private readonly ImageFactoryInterface $imageFactory, private readonly InsertTagParser $insertTags, private readonly string $projectDir, private readonly EntityCacheTags $cacheTags)
+    {
+    }
+
+    public function onFetchArticlesForFeed(FetchArticlesForFeedEvent $event): void
+    {
+        $pageModel = $event->getPageModel();
+        $archives = StringUtil::deserialize($pageModel->newsArchives, true);
+        $featured = match ($pageModel->feedFeatured) {
+            'featured' => true,
+            'unfeatured' => false,
+            default => null,
+        };
+
+        $newsModel = $this->framework->getAdapter(NewsModel::class);
+
+        $articles = $newsModel->findPublishedByPids($archives, $featured, $pageModel->maxFeedItems, 0, [
+            'return' => 'Array',
+        ]);
+
+        $event->setArticles($articles);
+    }
+
+    public function onTransformArticleForFeed(TransformArticleForFeedEvent $event): void
+    {
+        $article = $event->getArticle();
+
+        $item = new Item();
+
+        $item->setTitle($this->getTitle($article, $item, $event))
+            ->setLastModified($this->getLastModified($article, $item, $event))
+            ->setLink($this->getLink($article, $item, $event))
+            ->setPublicId($this->getPublicId($article, $item, $event))
+            ->setContent($this->getContent($article, $item, $event))
+        ;
+
+        $author = $this->getAuthor($article, $item, $event);
+
+        if ($author) {
+            $item->setAuthor($author);
+        }
+
+        $enclosures = $this->getEnclosures($article, $item, $event);
+
+        foreach ($enclosures as $enclosure) {
+            $item->addMedia($enclosure);
+        }
+
+        $event->setItem($item);
+    }
+
+    protected function getTitle(NewsModel $article, ItemInterface $item, TransformArticleForFeedEvent $event): string
```

Comment:
Why is it necessary to pass the `$item` and `$event` objects of none of them are used?

Reply:
Because it is `protected` and you may extend this class from your own listener to overwrite the methods.

Reply:
Since our event listeners are not part of the BC promise I’d go with `private` methods here.

Reply:
Me too. Extending listeners is kind of an anti-pattern.

Reply:
Agreed, however it saves you work when you only want to extend one or some methods. 
But in lack of a better idea and the fact that I just want to get this across the finish line, I made them all private in 9f9f381a7

Reply:
Removed unused arguments in [8344a47](https://github.com/contao/contao/pull/4682/commits/8344a47a174d419549371e79c7556a5f3267bfb5)